xclim/sdba/_adjustment.py

```diff
@@ -615,3 +615,390 @@ def extremes_adjust(
+    binX_sort = np.lexsort(binX[:, ::-1].T)
+    sorted_bins = binX[binX_sort]
+    _, binX_start, binX_count = np.unique(
+        sorted_bins, return_index=True, return_counts=True, axis=0
+    )
+    binX_start_sort = np.sort(binX_start)
+    binX_groups = np.split(binX_sort, binX_start_sort[1:])
+
+    out = np.empty(X.shape)
+    rng = np.random.default_rng()
+    # The plan row corresponding to a source bin indicates its probabilities to be transported to every target bin
+    for i, binX_group in enumerate(binX_groups):
+        # Get the plan row of this bin
+        pi = np.where((binX[binX_group[0]] == gridX).all(1))[0][0]
+        # Pick as much target bins for this source bin as there are points in the source bin
+        choice = rng.choice(range(muY.size), p=plan[pi, :], size=binX_count[i])
+        out[binX_group] = (gridY[choice] + 1 / 2) * bin_width + bin_origin
+
+    if spray_bins:
+        out += np.random.uniform(low=-bin_width / 2, high=bin_width / 2, size=out.shape)
+
+    return out
+
+
+@map_groups(scen=[Grouper.DIM])
+def otc_adjust(
+    ds: xr.Dataset,
+    dim: list,
+    pts_dim: str,
+    bin_width: list | None = None,
+    bin_origin: list | None = None,
+    num_iter_max: int | None = 100_000_000,
+    spray_bins: bool = True,
+    adapt_freq_thresh: dict | None = None,
+):
+    """Optimal Transport Correction of the bias of `hist` with respect to `ref`.
+
+    Parameters
+    ----------
+    ds : xr.Dataset
+        Dataset variables:
+            ref : training target
+            hist : training data
+    dim : list
+        The dimensions defining the distribution on which optimal transport is performed.
+    pts_dim : str
+        The dimension defining the multivariate components of the distribution.
+    bin_width : list | None
+        Bin widths for all dimensions.
+    bin_origin : list | None
+        Bin origins for all dimensions.
+    num_iter_max : int | None
+        Maximum number of iterations used in the earth mover distance algorithm.
+    spray_bins : bool = True
+        If `False`, output points are located at the center of their bin.
+        If `True`, a random location is picked uniformly inside their bin. Default is `True`.
+    adapt_freq_thresh : dict | None = None
+        Threshold for frequency adaptation per variable.
+
+    Returns
+    -------
+    xr.Dataset
+        Adjusted data
+    """
+    ref = ds.ref
+    hist = ds.hist
+
+    if adapt_freq_thresh is not None:
+        for var, thresh in adapt_freq_thresh.items():
+            if "units" not in ref.attrs["units"] or ref.attrs["units"] == "":
+                # Try to force units
+                units = str2pint(thresh).units
+                ref.attrs.update(units=str(units))
+            hist.loc[var] = _adapt_freq_hist(
+                xr.Dataset(
+                    {"ref": ref.sel({pts_dim: var}), "hist": hist.sel({pts_dim: var})}
+                ),
+                thresh,
+            )
```

Comment:
Forcing units could lead to errors (say you give `pr` in `mm/d` but the thresh is in `mm/s`). 

Adapted from my PR MBCn:

```python
# put this in adjustment.py
for var, thresh in adapt_freq_thresh.items():
    adapt_freq_thresh[var] = str(convert_units_to(thresh, self.train_units[var], context="hydro"))

# in _adjustment.py
for var, thresh in adapt_freq_thresh.items():
    hist.loc[var] = _adapt_freq_hist(
        xr.Dataset(
            {"ref": ref.sel({pts_dim: var}), "hist": hist.sel({pts_dim: var})}
        ),
        thresh,
    )
```
I can think I can push the branch MBCn today even if can get rid of some `tox` bugs that prevent me from merging, this will be necessary to have access to `train_units`

Reply:
The branch MBCn has been merged

Reply:
Thanks! Works like a charm